### PR TITLE
FEATURE: Provide central `PsrHttpFactory` that implements all interfaces from PSR 17

### PR DESCRIPTION
**Pull request by mficzel**
Injecting multiple factory objects to create PSR Http objects has been tedious in the past.

To overcome this the implementations of the PSR 17 methods are extracted into traits
and a central `PsrHttpFactory` is created. that implements all those interfaces.
That way a single object can be used to create Requests, Uris and Streams.

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)

### CODE REVIEW

**Comment by mficzel** [comment]
I think the __construct method in the ServerRequestFactoryTrait needs some additional thinking.

File: Neos.Http.Factories/Classes/HttpFactory.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+namespace Neos\Http\Factories;
+
+use Psr\Http\Message\RequestFactoryInterface;
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ServerRequestFactoryInterface;
+use Psr\Http\Message\StreamFactoryInterface;
+use Psr\Http\Message\UploadedFileFactoryInterface;
+use Psr\Http\Message\UriFactoryInterface;
+
+class HttpFactory implements ServerRequestFactoryInterface, RequestFactoryInterface, ResponseFactoryInterface, UriFactoryInterface, StreamFactoryInterface, UploadedFileFactoryInterface
```
Comment (kdambekalns) [inline Neos.Http.Factories/Classes/HttpFactory.php:13]: I think the name needs some more thinking. This factory doesn't "create HTTP"…

**Review by kdambekalns — COMMENTED** [review]
I get the idea. Not sure it's that much of a burden, but having a single factory for all those objects would be nice. 👍

Just the name `HttpFactory` seems off…

File: Neos.Http.Factories/Classes/HttpFactory.php (hunk 1 above)
Comment (mficzel) [reply Neos.Http.Factories/Classes/HttpFactory.php:13]: Agree ... suggestions are welcome aswell as an idea how to deal with the __constructor of the ServerRequestFactory

Comment (mficzel) [reply Neos.Http.Factories/Classes/HttpFactory.php:13]: Actually telling someone that the way to do a single http request is to inject 3 factories and plus Browser/Client (at least the engine is now injected by default) is an interesting experience.

Comment (kitsunet) [reply Neos.Http.Factories/Classes/HttpFactory.php:13]: yeha, I thought about that too, you can always sacrifice portability by directly instantiating a specific implementation that you then obvs need to require in your package.

Comment (kdambekalns) [reply Neos.Http.Factories/Classes/HttpFactory.php:13]: How is "supposed to work" according to "the spec"? Is there anything in the PSR or in Guzzle?

Comment (albe) [reply Neos.Http.Factories/Classes/HttpFactory.php:13]: > [](https://www.php-fig.org/psr/psr-17/#2-interfaces)2. Interfaces
The following interfaces MAY be implemented together within a single class or in separate classes.

https://www.php-fig.org/psr/psr-17/#2-interfaces

Comment (mficzel) [reply Neos.Http.Factories/Classes/HttpFactory.php:13]: The par looks to me weird but HttpObjectFactory sounds just fine

Comment (kitsunet) [reply Neos.Http.Factories/Classes/HttpFactory.php]: Either seem fine to me :) Although HttpObjectFactory sounds a bit smoother

Comment (kdambekalns) [reply Neos.Http.Factories/Classes/HttpFactory.php]: Indeed. That factory doesn't produce "HTTP", after all. ;)

Comment (mficzel) [reply Neos.Http.Factories/Classes/HttpFactory.php]: @kitsunet @kdambekalns renamed to HttpObjectFactory

**Review by kdambekalns — APPROVED** [review]

File: Neos.Http.Factories/Classes/HttpObjectFactory.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+namespace Neos\Http\Factories;
+
+use Psr\Http\Message\RequestFactoryInterface;
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ServerRequestFactoryInterface;
+use Psr\Http\Message\StreamFactoryInterface;
+use Psr\Http\Message\UploadedFileFactoryInterface;
+use Psr\Http\Message\UriFactoryInterface;
+
+class HttpObjectFactory implements ServerRequestFactoryInterface, RequestFactoryInterface, ResponseFactoryInterface, UriFactoryInterface, StreamFactoryInterface, UploadedFileFactoryInterface
```
Comment (mhsdesign) [inline Neos.Http.Factories/Classes/HttpObjectFactory.php]: One possible alternative name for this class could be PsrHttpFactory, as it clearly indicates that it implements the PSR HTTP message factories. Another option could be HttpMessageFactory, as it emphasizes the fact that this class is responsible for creating HTTP messages.

**Review by mhsdesign — CHANGES_REQUESTED** [review]
...

**Review by mhsdesign — APPROVED** [review]

**Review by ahaeslich — APPROVED** [review]

File: Neos.Http.Factories/Classes/ServerRequestFactoryTrait.php [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+    /**
+     * ServerRequestFactory constructor.
+     *
+     * @param UriFactoryInterface $uriFactory
+     * @param string $defaultUserAgent
+     * @param string $scriptPath
+     * @param string $defaultHttpVersion
+     */
+    public function __construct(
```
Comment (mhsdesign) [inline Neos.Http.Factories/Classes/ServerRequestFactoryTrait.php:46]: i just noticed that this trait comes with a constructor which is barely usable in real word ... and an odd piece of code to look at :D Maybe we should fix this somehow?

**Pull request merged**